### ISSUE TITLE
MFlux Optimised Quantization - MOQ

### ISSUE THREAD

**Issue by ianscrivener** [lpalbou/mlx-gen#13]
Laurent-Philippe, 
I took a very careful look through the way you do quantization for **mlx-gen**. 
- Hand-picked quantization is good, but that must have been quite time-consuming 
- Surprised that you quantize the VAE

Here's my report;

---

# mlx-gen quantization approach

Test case: **Z-Image Turbo**.

## Short answer

mlx-gen uses **one quantization method everywhere: MLX affine (integer) group quantization** via `mlx.nn.quantize`, at **4-bit or 8-bit**, **group size 64** (the MLX default).

There is **no** sensitivity measurement, **no** calibration data, **no** Hessian / GPTQ / AWQ, **no** SmoothQuant, and **no** Hadamard / rotation anywhere in the codebase. Where the word "sensitivity" appears (Wan, SwiftVR) it is a **hand-written comment describing a hard-coded layer allowlist** — a human decided which layers stay at higher precision; nothing is measured or tuned.

For **Z-Image Turbo**, quantization is completely uniform: the **same bit width is applied to the VAE, the transformer, and the text encoder**, driven by a single one-line rule.

---

### 1. Component declaration
`ZImageWeightDefinition.get_components()` declares three components — `vae`, `transformer`, `text_encoder` — and **none of them set `skip_quantization`**.

By contrast, SwiftVR, FIBO, Wan, SeedVR2, and Qwen-ControlNet all set `skip_quantization=True` on specific sub-models. Z-Image excludes nothing — **its VAE is quantized**.

### 2. The quantization predicate
`ZImageWeightDefinition.quantization_predicate` is the entire policy:

```python
@staticmethod
def quantization_predicate(path: str, module) -> bool:
    return hasattr(module, "to_quantized")
```

Every quantizable module (all `nn.Linear`, etc.) in all three components is quantized to the requested bit width. No per-layer variation, no path checks, no shape checks.

### 3. Applying it
`WeightApplier.apply_and_quantize` (`src/mflux/models/common/weights/loading/weight_applier.py`) drives all three models at once:

- **`QuantizationResolution.resolve(stored, requested)`** picks the bit width:
  | stored in checkpoint | `-q` flag | result |
  |---|---|---|
  | none | none | stay full precision (bf16) |
  | none | `N` | quantize on the fly to `N` bits |
  | `M` | none | use `M` bits |
  | `M` | `N` | use `M` bits; warn and ignore `-q N` if they differ |
- Then `nn.quantize(model, class_predicate=<predicate>, bits=bits)` is called on each of `vae`, `transformer`, `text_encoder`.
- `nn.quantize` is called **without `group_size`** → MLX default **64**. Z-Image has no guard for dimensions not divisible by 64 (Qwen does); Z-Image tensors are assumed to divide cleanly.

### 4. Pre-quantized checkpoints
The recommended path (`filipstrand/Z-Image-Turbo-mflux-4bit`) is a checkpoint already quantized by this same code. On load, the runtime module skeleton is quantized *first* to match the stored layout, then the stored weights are copied in.

---

## Approach comparison: VAE vs. text encoder vs. transformer (Z-Image Turbo)

| Aspect | VAE | Text encoder | Transformer |
|---|---|---|---|
| Quantization method | MLX affine int group quant (`nn.quantize`) | same | same |
| Bit widths | 4 or 8 (run-wide) | 4 or 8 (run-wide) | 4 or 8 (run-wide) |
| Group size | 64 (MLX default) | 64 | 64 |
| Modules quantized | all with `to_quantized` | all with `to_quantized` | all with `to_quantized` |
| Per-layer / mixed precision | none — uniform | none — uniform | none — uniform |
| Sensitivity analysis / tuning | none | none | none |
| Calibration data | none | none | none |
| Hadamard / rotation / SmoothQuant / GPTQ / AWQ | none | none | none |
| Layers held at higher precision | none | none | none |
| Skipped from quantization | nothing (`skip_quantization` not set) | nothing | nothing |
| Precision when un-quantized | `ModelConfig.precision` (bf16) | bf16 | bf16 |

All three rows are identical. There is no differentiation between components for Z-Image.

---

## For contrast: the one model with mixed precision (Qwen-Image)

Z-Image is the simple case. **Qwen-Image** is the only model in the repo with mixed precision — and it is still **not measured**, it is a **hand-curated hard-coded allowlist**:

- `QwenWeightDefinition._is_bf16_q4_sensitive_transformer_path` lists specific paths by name: `img_in`, `txt_in`, `norm_out.linear`, `proj_out`, `time_text_embed.*`, `.img_mod_linear`, `.txt_mod_linear`.
- When the user requests 4-bit, the predicate returns `{"bits": 8}` (MLX's per-module override) for those paths, or skips them, keeping the rest at 4-bit.
- `_quantized_linear_bits` back-computes bits from tensor shapes (`weight.shape[-1] * 32 / (scales.shape[-1] * 64)`) — this only *reads* what precision an existing checkpoint used so the runtime skeleton matches; it is not analysis.
- `_can_quantize_group64` skips any layer whose last dim is not divisible by 64.
- The Qwen text encoder gets its own predicate: vision-tower layers forced to 8-bit under a 4-bit request, group-64-incompatible layers skipped.

Even this "advanced" path is: a person picked sensitive layers from experience and wrote them into the source. No automated sensitivity metric, no calibration set, no rotation.

---

## Per-model VAE quantization (repo-wide)

| Model | VAE / decoder quantized? |
|---|---|
| Z-Image | **Yes** (no `skip_quantization`) |
| Qwen-Image | Yes |
| FIBO | No (`skip_quantization=True`) |
| Wan | No (`skip_quantization=True`) |
| SwiftVR | No (ReAE excluded, `skip_quantization=True`) |
| SeedVR2 | No (`skip_quantization=True`) |

---

## Key files

- `src/mflux/models/z_image/weights/z_image_weight_definition.py` — the one-line Z-Image predicate; components list (no skip)
- `src/mflux/models/z_image/z_image_initializer.py` — `_apply_weights` passes vae/transformer/text_encoder together
- `src/mflux/models/common/weights/loading/weight_applier.py` — `apply_and_quantize`, calls `nn.quantize`
- `src/mflux/models/common/resolution/quantization_resolution.py` — stored-vs-requested bit-width rules
- `src/mflux/models/qwen/weights/qwen_weight_definition.py` — the hand-coded mixed-precision allowlist (contrast)
